### PR TITLE
Add Ty/LifetimeOutlives GenericPredicate variants

### PR DESCRIPTION
**Pull request by Veykril**

### CODE REVIEW

**Comment by Veykril** [comment]
Should've said that I'm working on this part in Zulip to not clash, but I kinda forgot sorry. Hope you didn't work on anything in that area Florian 😅

**Comment by flodiebold** [comment]
Not yet! I'll continue working on RA later today / on the weekend though.

File: crates/hir_ty/src/lib.rs [hunk 1]
```diff
@@ -561,6 +559,34 @@ impl TypeWalk for TraitRef {
...
+    ) {
+        self.ty.walk_mut_binders(f, binders);
+    }
+}
+
+impl TypeWalk for LifetimeOutlives {
+    fn walk(&self, _: &mut impl FnMut(&Ty)) {}
+
+    fn walk_mut_binders(&mut self, _: &mut impl FnMut(&mut Ty, DebruijnIndex), _: DebruijnIndex) {}
+}
```
Comment (Veykril) [inline crates/hir_ty/src/lib.rs:588]: Our `TypeWalk` is basically a very weak version of chalks `Fold` right? Hence I did this implementation so this works with `do_canonicalize`

**Comment by flodiebold** [comment]
I'm not really sure why you're adding these variants though -- we're not going to be constructing them anyway, right?

**Comment by Veykril** [comment]
Isn't our GenericPredicate chalk's WhereClause? I figured moving them closer would help. Might have misunderstood this part

**Comment by flodiebold** [comment]
Yes (and I think renaming GenericPredicate to WhereClause is a step to take). But as long as we're never constructing these variants, there's not much use in having them, right? I guess it would force us to add them to a few matches. The point of these changes is to change the code in the same way we would need to change it when switching to the Chalk versions, but if we're never using them it doesn't make a difference.

**Comment by Veykril** [comment]
Hmm ye that makes sense, guess I was a bit overeager here 👍

**Comment by flodiebold** [comment]
OTOH the HirDisplay additions are going to be something we'll need...

**Comment by Veykril** [comment]
Well, the predicate displays can be pulled out of this PR when needed. The lifetime stuff I could pull into a new PR as those won't give us more work when switching since that is already basically all chalk.

**Pull request closed without merging**